runtime/lua/vim/fs.lua

```diff
@@ -334,30 +334,147 @@ function M.find(names, opts)
   return matches
 end
 
+--- Split a Windows path into a prefix and a body, such that the body can be processed like a POSIX
+--- path. The path must use forward slashes as path separator.
+---
+--- Does not check if the path is a valid Windows path. Invalid paths will give invalid results.
+---
+--- Examples:
+--- - `//./C:/foo/bar` -> `//./C:`, `/foo/bar`
+--- - `//?/UNC/server/share/foo/bar` -> `//?/UNC/server/share`, `/foo/bar`
+--- - `//./system07/C$/foo/bar` -> `//./system07`, `/C$/foo/bar`
+--- - `C:/foo/bar` -> `C:`, `/foo/bar`
+--- - `C:foo/bar` -> `C:`, `foo/bar`
+---
+--- @param path string Path to split.
+--- @return string, string : prefix, body
+local function split_windows_path(path)
+  local prefix = ''
+
+  --- Match pattern. If there is a match, move the matched pattern from the path to the prefix.
+  --- Returns the matched pattern.
+  ---
+  --- @param pattern string Pattern to match.
+  --- @param nomatch_error string? Error message if the pattern doesn't match.
+  --- @return string|nil Matched pattern
+  local function match_to_prefix(pattern, nomatch_error)
+    local match = path:match(pattern)
+
+    if match then
+      prefix = prefix .. match --[[ @as string ]]
+      path = path:sub(#match + 1)
+    elseif nomatch_error then
+      error(nomatch_error)
+    end
+
+    return match
+  end
+
+  local function process_unc_path()
+    match_to_prefix('[^/]+/+[^/]+/+', 'Invalid Windows UNC path')
```

Comment:
can `match_to_prefix` and `process_unc_path` be reworked to return multiple values instead of mutating `prefix` (which is subtle and generally should be avoided).

usually a good idea to also print the arg at the end of the error.

```suggestion
    _, prefix = match_to_prefix('[^/]+/+[^/]+/+', ('Invalid Windows UNC path: "%s"'):format(...))
```